src/wavm.cpp

```diff
@@ -0,0 +1,183 @@
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "getCallDataSize", U32, getCallDataSize)
+  {
+    printf("getCallDataSize\n");
+    U32 ret = interface.top()->eeiGetCallDataSize();
+    return ret;
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "getCodeSize", U32, getCodeSize)
+  {
+    printf("getCodeSize\n");
+    U32 ret = interface.top()->eeiGetCodeSize();;
+    return ret;
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "call", U32, call, I64 gas, U32 addressOffset, U32 valueOffset, U32 dataOffset, U32 dataLength)
+  {
+    printf("call  gas=%lu  addressOffset=%i  valueOffset=%i  dataOffset=%i  dataLength=%i\n", gas, addressOffset, valueOffset, dataOffset, dataLength);
+    U32 result = interface.top()->eeiCall(hera::EthereumInterface::EEICallKind::Call, gas, addressOffset, valueOffset, dataOffset, dataLength);
+    return result;
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION_WITH_MEM_AND_TABLE(ethereum, "useGas", void, useGas, I64 amount)
+  {
+    printf("useGas  amount=%lu\n",amount);
+    interface.top()->eeiUseGas(amount);
+  }
+
+
+  // this is needed for resolving names of imported host functions
+  struct MyResolver : Runtime::Resolver {
+    Runtime::Compartment* compartment;
+    HashMap<std::string, Runtime::ModuleInstance*> moduleNameToInstanceMap;
+
+    MyResolver(Runtime::Compartment* inCompartment) : compartment(inCompartment) {}
+
+    bool resolve(const std::string& moduleName,
+           const std::string& exportName,
+           IR::ObjectType type,
+           Runtime::Object*& outObject) override {
+      auto namedInstance = moduleNameToInstanceMap.get(moduleName);
+      outObject = getInstanceExport(*namedInstance, exportName);
+      return true;
+    }
+  };
+
+} //namespace wavm_host_module
+
+
+
+ExecutionResult WavmEngine::execute(
+  evmc_context* context,
+  vector<uint8_t> const& code,
+  vector<uint8_t> const& state_code,
+  evmc_message const& msg,
+  bool meterInterfaceGas
+) {
+  HERA_DEBUG << "Executing with wavm...\n";
+
+  // set up a new ethereum interface just for this contract invokaction
+  ExecutionResult result;
+  wavm_host_module::interface.push(new WavmEthereumInterface{context, state_code, msg, result, meterInterfaceGas});
+
+  // first parse module
+  std::vector<U8> codeBytes = static_cast< std::vector<U8> > (code);
+  IR::Module moduleAST; 
+  loadBinaryModule(codeBytes.data(), codeBytes.size(), moduleAST);
+
+  // next set up the host module. Note: in ewasm, we create a new VM for each call to a module, so we must instantiate a new host module for each of these VMs, this is inefficient, but OK for prototyping.
+  // compartment is like the Wasm store, represents the VM, has lists of globals, memories, tables, and also has wavm's runtime stuff
+  Runtime::Compartment* compartment = Runtime::createCompartment();
+  // context stores the compartment and some other stuff
+  Runtime::Context* wavm_context = Runtime::createContext(compartment);
```

Comment:
Same here, wavm don't seem to manage this pointer. However it loosely references `compartment`...